engine/table/src/main/java/io/deephaven/engine/table/impl/locations/impl/AbstractTableLocationProvider.java

```diff
@@ -308,38 +408,46 @@ protected final void handleTableLocationKeyRemoved(@NotNull final TableLocationK
      * necessary. See {@link #removeTableLocationKey(TableLocationKey)} for additional discussions of semantics.
      * 
      * @param locationKey the TableLocation that was removed
-     * @param transactionToken The token identifying the transaction
+     * @param transactionToken The token identifying the transaction (or null if not part of a transaction)
      */
     protected void handleTableLocationKeyRemoved(
             @NotNull final TableLocationKey locationKey,
             @Nullable final Object transactionToken) {
         if (!supportsSubscriptions()) {
-            maybeClearLocationForRemoval(tableLocations.remove(locationKey));
             return;
         }
 
+        // When removing a location in a transaction, check for logical consistency.
+        // 1. If the location was already removed in this transaction, we have a problem. A transaction should not
+        // remove the same location twice.
+        // 2. If the location was already added in this transaction, we have a problem. A transaction should not
+        // add then remove the same location.
+        if (transactionToken != null) {
+            synchronized (transactionLock) {
+                final Transaction transaction = transactions.get(transactionToken);
+                if (transaction == null) {
+                    throw new IllegalStateException(
+                            "No transaction with token " + transactionToken + " is currently open.");
+                }
+                transaction.removeLocationKey(locationKey);
+                return;
+            }
+        }
+
+        // If we're not in a transaction, we should push this key immediately.
         synchronized (subscriptions) {
             final Object removedLocation = tableLocations.remove(locationKey);
             if (removedLocation != null) {
-                maybeClearLocationForRemoval(removedLocation);
                 if (subscriptions.deliverNotification(
                         Listener::handleTableLocationKeyRemoved,
                         locationKey.makeImmutable(),
-                        transactionToken,
                         true)) {
                     onEmpty();
                 }
             }
         }
     }
 
-    private void maybeClearLocationForRemoval(@Nullable final Object removedLocation) {
```

Comment:
I think you need to hold a ref count on the TL as long as it's in your map. (That is, if you add an actual TL, if it's an aTL, bump the ref count. When you remove an actual TL, if it's an aTL, dec the ref count. No need to worry about TLKs.)